<a href="https://colab.research.google.com/github/SaiArjunSairamje/Python_Libraries-Scaler/blob/main/Pandas_2%20(Dataset%3A%20IMDB_Director_Insights)%20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Importing Libraries and Dataset "McKinsey LifeExpentancy Insights" (Cont.)**
***

In [70]:
import numpy as np
import pandas as pd

In [71]:
!wget "https://drive.google.com/uc?export=download&id=1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_" -O mckinsey.csv

--2023-07-13 15:18:23--  https://drive.google.com/uc?export=download&id=1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_
Resolving drive.google.com (drive.google.com)... 142.251.111.138, 142.251.111.113, 142.251.111.139, ...
Connecting to drive.google.com (drive.google.com)|142.251.111.138|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-0s-68-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/ilfv3ghjbbklpp6ab7m2r8s4nse0j55i/1689261450000/14302370361230157278/*/1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_?e=download&uuid=8527882b-d567-4fcc-ade1-1e4faccbe745 [following]
--2023-07-13 15:18:23--  https://doc-0s-68-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/ilfv3ghjbbklpp6ab7m2r8s4nse0j55i/1689261450000/14302370361230157278/*/1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_?e=download&uuid=8527882b-d567-4fcc-ade1-1e4faccbe745
Resolving doc-0s-68-docs.googleusercontent.com (doc-0s-68-docs.googleusercontent.com)... 142.251.111.132, 

In [72]:
df = pd.read_csv("mckinsey.csv")

In [73]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

# **Duplicate Rows based data**
***

**`NOTE:`**            
> **df.index** attribute in pandas DataFrame returns the index of the DataFrame, which represents the row labels or the row indices.  

> **len(df.index)** returns the length or the number of elements in the index of a DataFrame df. It represents the total number of rows in the DataFrame.  

> **df.loc** accessor in pandas DataFrame is used for label-based indexing. It allows you to access specific rows and columns of a DataFrame using labels or boolean masks.

> The code **df.loc[len(df.index)]** is used to access the last row of a DataFrame **df**. It uses the **len(df.index)** expression to retrieve the index label corresponding to the last row and then accesses that row using **df.loc**.



### **Let's add some duplicate data**

In [74]:
df.loc[len(df.index)] = ['India',2000,13500000,"asia",37.08,900.23]
df.loc[len(df.index)] = ['Sri Lanka',2022 ,130000000,"asia",80.00,500.00]
df.loc[len(df.index)] = ['Sri Lanka',2022 ,130000000,"asia",80.00,500.00]
df.loc[len(df.index)] = ['India',2000 ,13500000,"asia",80.00,900.23]
df.tail(10)

country  year  population continent  life_exp     gdp_cap
1698   Zimbabwe  1982     7636524    Africa    60.363  788.855041
1699   Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700   Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701   Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702   Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703   Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704      India  2000    13500000      asia    37.080  900.230000
1705  Sri Lanka  2022   130000000      asia    80.000  500.000000
1706  Sri Lanka  2022   130000000      asia    80.000  500.000000
1707      India  2000    13500000      asia    80.000  900.230000

### **How to find duplicate rows in Pandas ???**

The **df.duplicated()** method in Pandas is used to identify duplicate rows in a DataFrame. It returns a Boolean Series that indicates whether each row is a duplicate or not.

The resulting Boolean Series indicates **False** for non-duplicate rows and **True** for the duplicate row(s).


In [75]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1703    False
1704    False
1705    False
1706     True
1707    False
Length: 1708, dtype: bool

### **How to extract duplicate rows ???**

> **`Approach (1):`** Normal `indexing` method with **`duplicated()`** function

> **`Approach (2):`** Using the **`loc`** indexer with **`duplicated()`** function

To extract the duplicate rows using the **loc** indexer, we use **df.loc[duplicates]**, where **duplicates** is the boolean Series that indicates which rows are duplicates. The resulting DataFrame **duplicate_rows** contains only the rows that are duplicates.



In [76]:
# Approach (1)

df[df.duplicated()]

country  year  population continent  life_exp  gdp_cap
1706  Sri Lanka  2022   130000000      asia      80.0    500.0

In [77]:
# Approach (2)

df.loc[df.duplicated()]

country  year  population continent  life_exp  gdp_cap
1706  Sri Lanka  2022   130000000      asia      80.0    500.0

**Why we can't extract duplicated rows using iloc() ???**

The **iloc()** function in Pandas is primarily used for integer-based indexing and slicing of rows and columns in a DataFrame. It allows you to access specific rows and columns based on their integer positions.

While **iloc()** can be used to select rows based on their positions, it does not have a direct functionality to identify and extract duplicated rows. This is because duplicate rows are not necessarily consecutive in the DataFrame, and their positions may vary based on the sorting or filtering criteria applied to the data.

In [78]:
df.iloc[df.duplicated()]  # this will throw an error message

NotImplementedError: ignored

### **How to remove the duplicate rows ???**

`Note` that **df.loc[df.duplicated()]** returns a new DataFrame that includes only the duplicate rows. If you want to retain the original DataFrame and remove the duplicate rows, you can use **df.drop_duplicates()** as shown in the previous response.

In [79]:
df.tail()

country  year  population continent  life_exp     gdp_cap
1703   Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704      India  2000    13500000      asia    37.080  900.230000
1705  Sri Lanka  2022   130000000      asia    80.000  500.000000
1706  Sri Lanka  2022   130000000      asia    80.000  500.000000
1707      India  2000    13500000      asia    80.000  900.230000

In [80]:
df.drop_duplicates()

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      asia    37.080  900.230000
1705    Sri Lanka  2022   130000000      asia    80.000  500.000000
1707        India  2000    13500000      asia    80.000  900.230000

[1707 rows x 6 columns]

In [81]:
df.tail()  # the original dataFrame still has the duplicate row and we need to fix it...

country  year  population continent  life_exp     gdp_cap
1703   Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704      India  2000    13500000      asia    37.080  900.230000
1705  Sri Lanka  2022   130000000      asia    80.000  500.000000
1706  Sri Lanka  2022   130000000      asia    80.000  500.000000
1707      India  2000    13500000      asia    80.000  900.230000

**Since there are 2 rows having same values, Which duplicate row to be removed ???**

The **drop_duplicates()** method in Pandas is used to remove duplicate rows from a DataFrame. It provides the **keep** parameter to control which duplicate values to keep. There are 2 primary ways to use **drop_duplicates()** with different values for the **keep** parameter:

> **`Approach (1):` keep='first'** (default)

> **`Approach (2):` keep='last'**

**Approach (1): keep='first'** (default)         
By default, the **drop_duplicates()** method keeps the first occurrence of each duplicated row and removes the subsequent occurrences.

In [82]:
df.drop_duplicates(keep = "first")

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      asia    37.080  900.230000
1705    Sri Lanka  2022   130000000      asia    80.000  500.000000
1707        India  2000    13500000      asia    80.000  900.230000

[1707 rows x 6 columns]

**Approach (2): keep='last'**            
Setting **keep='last'** as the parameter value for **drop_duplicates()** keeps the last occurrence of each duplicated row and removes the previous occurrences.

In [83]:
df.drop_duplicates(keep = "last")

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      asia    37.080  900.230000
1706    Sri Lanka  2022   130000000      asia    80.000  500.000000
1707        India  2000    13500000      asia    80.000  900.230000

[1707 rows x 6 columns]

# **Starting fresh for new topic**
***

In [84]:
df = pd.read_csv("mckinsey.csv")

In [85]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

### **Which among `loc()` & `iloc()` takes the last index as well while slicing ???**

In [86]:
# loc (i.e. explicit index)

df.loc[1:4]

country  year  population continent  life_exp     gdp_cap
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [87]:
df.iloc[1:4]

country  year  population continent  life_exp     gdp_cap
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138

### **Since we know how to slice Rows, how can we slice Cloumns ???**

To slice columns in a Pandas DataFrame, you can use **indexing** or the **loc** and **iloc** indexers.

> **`Approach (1):` Using Indexing**        
You can slice columns by specifying the column names as a list within the indexing operator **[]**

> **`Approach (2):` Using loc indexer**        
The **loc** indexer allows you to slice columns by label-based indexing. You can provide column labels as a list or a slice object within the **loc** indexer.

> **`Approach (3):` Using iloc indexer**     
The **iloc** indexer allows you to slice columns by positional indexing. You can provide column positions as a list or a slice object within the **iloc** indexer.

In [88]:
df.head(7)

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106
5  Afghanistan  1977    14880372      Asia    38.438  786.113360
6  Afghanistan  1982    12881816      Asia    39.854  978.011439

In [89]:
df.iloc[1:5, 1:4]  # Using iloc indexer - Approach (2)

year  population continent
1  1957     9240934      Asia
2  1962    10267083      Asia
3  1967    11537966      Asia
4  1972    13079460      Asia

In [90]:
df.loc[1:5, 1:4] # Using loc indexer - Approach (3) trial
# we will get error since we do have any such columns with numbers

TypeError: ignored

In [91]:
df.loc[1:5, ["country", "life_exp"]] # Using Indexing - Approach (1) with Approach (3) method

country  life_exp
1  Afghanistan    30.332
2  Afghanistan    31.997
3  Afghanistan    34.020
4  Afghanistan    36.088
5  Afghanistan    38.438

In [92]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [93]:
df.loc[1:5, "year" : "continent"] # using loc indexer - Approach (3)

year  population continent
1  1957     9240934      Asia
2  1962    10267083      Asia
3  1967    11537966      Asia
4  1972    13079460      Asia
5  1977    14880372      Asia

In [94]:
df.iloc[1:5, "year" : "continent"]

TypeError: ignored

In [95]:
df.head(20)

country  year  population continent  life_exp      gdp_cap
0   Afghanistan  1952     8425333      Asia    28.801   779.445314
1   Afghanistan  1957     9240934      Asia    30.332   820.853030
2   Afghanistan  1962    10267083      Asia    31.997   853.100710
3   Afghanistan  1967    11537966      Asia    34.020   836.197138
4   Afghanistan  1972    13079460      Asia    36.088   739.981106
5   Afghanistan  1977    14880372      Asia    38.438   786.113360
6   Afghanistan  1982    12881816      Asia    39.854   978.011439
7   Afghanistan  1987    13867957      Asia    40.822   852.395945
8   Afghanistan  1992    16317921      Asia    41.674   649.341395
9   Afghanistan  1997    22227415      Asia    41.763   635.341351
10  Afghanistan  2002    25268405      Asia    42.129   726.734055
11  Afghanistan  2007    31889923      Asia    43.828   974.580338
12      Albania  1952     1282697    Europe    55.230  1601.056136
13      Albania  1957     1476505    Europe    59.280  1942.284244
14      Albania  1962     1728137    Europe    64.820  2312.888958
15      Albania  1967     1984060    Europe    66.220  2760.196931
16      Albania  1972     2263554    Europe    67.690  3313.422188
17      Albania  1977     2509048    Europe    68.930  3533.003910
18      Albania  1982     2780097    Europe    70.420  3630.880722
19      Albania  1987     3075321    Europe    72.000  3738.932735

### **Slicing with jump statement for `iloc()` and `loc()`**

In [96]:
# How many Rows & Columns will we get as o/p ???

df.iloc[ [0,10, 20, 100] , [0, 2, 3] ]  # Multi-Indexing approach

country  population continent
0    Afghanistan     8425333      Asia
10   Afghanistan    25268405      Asia
20       Albania     3326498    Europe
100   Bangladesh    70759295      Asia

In [97]:
df.iloc[1:10:2] # Slicing with jump statement

country  year  population continent  life_exp     gdp_cap
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
5  Afghanistan  1977    14880372      Asia    38.438  786.113360
7  Afghanistan  1987    13867957      Asia    40.822  852.395945
9  Afghanistan  1997    22227415      Asia    41.763  635.341351

In [98]:
df.loc[1:10:2] # Slicing with jump statement with loc also works but if it had a,b,c,d etc. indeces, it won't work

country  year  population continent  life_exp     gdp_cap
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
5  Afghanistan  1977    14880372      Asia    38.438  786.113360
7  Afghanistan  1987    13867957      Asia    40.822  852.395945
9  Afghanistan  1997    22227415      Asia    41.763  635.341351

# **Pandas built-in operations**
***
Pandas provides a wide range of built-in operations that can be performed on **DataFrames** and **Series**. Here are some commonly used built-in operations in Pandas:
**(1)** Descriptive Statistics **(2)** Data Manipulation **(3)** Data Combination or Data Merging Operations

### **(1) Descriptive Statistics**

**df.mean()**: Compute the mean value of each column.           
**df.sum()**: Calculates the sum values of each column.            
**df.count()**: Counts the non-missing values of each column.      
**df.median()**: Compute the median value of each column.          
**df.min()**: Find the minimum value in each column.           
**df.max()**: Find the maximum value in each column.          
**df.std()**: Calculate the standard deviation of each column.           
**df.describe()**: Generate descriptive statistics summary of the DataFrame.

In [99]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [100]:
le = df["life_exp"]
print(le)

0       28.801
1       30.332
2       31.997
3       34.020
4       36.088
         ...  
1699    62.351
1700    60.377
1701    46.809
1702    39.989
1703    43.487
Name: life_exp, Length: 1704, dtype: float64


In [101]:
le.mean()

59.474439366197174

In [102]:
le.sum()

101344.44467999999

In [103]:
le.count()

1704

In [104]:
le.sum()/le.count()  # average value same as mean value

59.474439366197174

In [105]:
le.min()

23.599

In [106]:
le.max()

82.603

In [107]:
le.std()

12.917107415241192

The **describe()** function in Pandas is used to generate descriptive statistics of a DataFrame. When called on a DataFrame df, it provides a summary of the central tendency, dispersion, and shape of the distribution of the numerical columns in the DataFrame.

The **describe()** function provides the following statistics for each numeric column:

* **Count**: The number of non-null values in the column.     
* **Mean**: The average value of the column.        
* **Standard Deviation**: The measure of the spread of values around the mean.           
* **Minimum**: The minimum value in the column.        
* **25th Percentile**: The value below which 25% of the data falls.           
* **50th Percentile (Median)**: The middle value in the column.    
* **75th Percentile**: The value below which 75% of the data falls.       
* **Maximum**: The maximum value in the column.

In [108]:
le.describe()

count    1704.000000
mean       59.474439
std        12.917107
min        23.599000
25%        48.198000
50%        60.712500
75%        70.845500
max        82.603000
Name: life_exp, dtype: float64

**`NOTE:`**        
By default, **describe()** only provides the summary for numeric columns. However, you can include all columns by passing **include='all'** as an argument: **df.describe(include='all')**.

**`Example:`**       
By using **df.describe(include='all')**, the summary statistics are provided for all columns, including both numerical and object (categorical) columns. The output shows statistics such as count, unique values, top value, frequency, mean, standard deviation, minimum value, 25th percentile, median (50th percentile), 75th percentile, and maximum value for each column.


In [109]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [110]:
df[["country", "year", "population"]].describe(include = 'all')

country        year    population
count          1704  1704.00000  1.704000e+03
unique          142         NaN           NaN
top     Afghanistan         NaN           NaN
freq             12         NaN           NaN
mean            NaN  1979.50000  2.960121e+07
std             NaN    17.26533  1.061579e+08
min             NaN  1952.00000  6.001100e+04
25%             NaN  1965.75000  2.793664e+06
50%             NaN  1979.50000  7.023596e+06
75%             NaN  1993.25000  1.958522e+07
max             NaN  2007.00000  1.318683e+09

### **(2) Data Manipulation**

**df.drop()**: Drop specified rows or columns from the DataFrame.      
**df.rename()**: Rename columns or index labels.          
**df.sort_values()**: Sort the DataFrame by one or more columns. (`NOTE:` Similar to **"Order By"** in **SQL**)      

In [111]:
df.sort_values(['life_exp'])  # deafult order in Ascending similar to Order By in SQL

country  year  population continent  life_exp       gdp_cap
1292            Rwanda  1992     7290203    Africa    23.599    737.068595
0          Afghanistan  1952     8425333      Asia    28.801    779.445314
552             Gambia  1952      284320    Africa    30.000    485.230659
36              Angola  1952     4232095    Africa    30.015   3520.610273
1344      Sierra Leone  1952     2143249    Africa    30.331    879.787736
...                ...   ...         ...       ...       ...           ...
1487       Switzerland  2007     7554661    Europe    81.701  37506.419070
695            Iceland  2007      301931    Europe    81.757  36180.789190
802              Japan  2002   127065841      Asia    82.000  28604.591900
671   Hong Kong, China  2007     6980412      Asia    82.208  39724.978670
803              Japan  2007   127467972      Asia    82.603  31656.068060

[1704 rows x 6 columns]

In [112]:
df.sort_values(['life_exp'], ascending = False)  # Sorting in Descending order

country  year  population continent  life_exp       gdp_cap
803              Japan  2007   127467972      Asia    82.603  31656.068060
671   Hong Kong, China  2007     6980412      Asia    82.208  39724.978670
802              Japan  2002   127065841      Asia    82.000  28604.591900
695            Iceland  2007      301931    Europe    81.757  36180.789190
1487       Switzerland  2007     7554661    Europe    81.701  37506.419070
...                ...   ...         ...       ...       ...           ...
1344      Sierra Leone  1952     2143249    Africa    30.331    879.787736
36              Angola  1952     4232095    Africa    30.015   3520.610273
552             Gambia  1952      284320    Africa    30.000    485.230659
0          Afghanistan  1952     8425333      Asia    28.801    779.445314
1292            Rwanda  1992     7290203    Africa    23.599    737.068595

[1704 rows x 6 columns]

In [113]:
df.sort_values(['year', 'life_exp']) # Sorting by 2 columns. By default both are Ascending columns.

country  year  population continent  life_exp       gdp_cap
0          Afghanistan  1952     8425333      Asia    28.801    779.445314
552             Gambia  1952      284320    Africa    30.000    485.230659
36              Angola  1952     4232095    Africa    30.015   3520.610273
1344      Sierra Leone  1952     2143249    Africa    30.331    879.787736
1032        Mozambique  1952     6446316    Africa    31.286    468.526038
...                ...   ...         ...       ...       ...           ...
71           Australia  2007    20434176   Oceania    81.235  34435.367440
1487       Switzerland  2007     7554661    Europe    81.701  37506.419070
695            Iceland  2007      301931    Europe    81.757  36180.789190
671   Hong Kong, China  2007     6980412      Asia    82.208  39724.978670
803              Japan  2007   127467972      Asia    82.603  31656.068060

[1704 rows x 6 columns]

In [114]:
df.sort_values(['year', 'life_exp'], ascending = [True, False])  # Sorting by 2 columns in which 1 in Ascending & other in Descending

country  year  population continent  life_exp       gdp_cap
1140        Norway  1952     3327728    Europe    72.670  10095.421720
684        Iceland  1952      147962    Europe    72.490   7267.688428
1080   Netherlands  1952    10381988    Europe    72.130   8941.571858
1464        Sweden  1952     7124673    Europe    71.860   8527.844662
408        Denmark  1952     4334000    Europe    70.780   9692.385245
...            ...   ...         ...       ...       ...           ...
887        Lesotho  2007     2012649    Africa    42.592   1569.331442
1355  Sierra Leone  2007     6144562    Africa    42.568    862.540756
1691        Zambia  2007    11746035    Africa    42.384   1271.211593
1043    Mozambique  2007    19951656    Africa    42.082    823.685621
1463     Swaziland  2007     1133066    Africa    39.613   4513.480643

[1704 rows x 6 columns]

### **(3) Data Combination or Data Merging Operations**

*   **pd.concat()**
*   **pd.merge()**
*   **pd.join()** (**`NOTE:`** Additional topic external from **Scaler**)

**pd.concat()**:        

This function is used to concatenate two or more DataFrames along a particular axis. It can be used to combine DataFrames vertically (along rows) or horizontally (along columns).

In [115]:
users = pd.DataFrame({"userid":[1, 2, 3], "name":["AJ", "Manoj", "Mukesh"]})
users

userid    name
0       1      AJ
1       2   Manoj
2       3  Mukesh

In [116]:
msgs = pd.DataFrame({"userid":[1, 1, 2, 4], "msg":['hmm', "acha", "theek hai", "nice"]})
msgs

userid        msg
0       1        hmm
1       1       acha
2       2  theek hai
3       4       nice

In [117]:
pd.concat([users, msgs], axis = 0) # Vertical Stack - little bit similar to Unions in SQL

userid    name        msg
0       1      AJ        NaN
1       2   Manoj        NaN
2       3  Mukesh        NaN
0       1     NaN        hmm
1       1     NaN       acha
2       2     NaN  theek hai
3       4     NaN       nice

In [118]:
pd.concat([users, msgs], axis = 0, ignore_index = True)

userid    name        msg
0       1      AJ        NaN
1       2   Manoj        NaN
2       3  Mukesh        NaN
3       1     NaN        hmm
4       1     NaN       acha
5       2     NaN  theek hai
6       4     NaN       nice

In [119]:
pd.concat([users, msgs], axis = 1) # Horizonal Stack
# NOTE: When ever there is a "NaN" value in a numerical column, all values will become "Float" datatype

userid    name  userid        msg
0     1.0      AJ       1        hmm
1     2.0   Manoj       1       acha
2     3.0  Mukesh       2  theek hai
3     NaN     NaN       4       nice

**pd.merge()**:    
      
This function is used to merge two DataFrames based on common columns or indices. It performs database-style joins (e.g., inner join, outer join, etc.) to combine DataFrames.

In [120]:
users

userid    name
0       1      AJ
1       2   Manoj
2       3  Mukesh

In [121]:
msgs

userid        msg
0       1        hmm
1       1       acha
2       2  theek hai
3       4       nice

In [122]:
users.merge(msgs, on = "userid") # Inner Join

userid   name        msg
0       1     AJ        hmm
1       1     AJ       acha
2       2  Manoj  theek hai

In [123]:
# How to perform particular type of Join ???

users.merge(msgs, on = "userid" , how = "outer") # Outer Join

userid    name        msg
0       1      AJ        hmm
1       1      AJ       acha
2       2   Manoj  theek hai
3       3  Mukesh        NaN
4       4     NaN       nice

In [124]:
users.merge(msgs, on = "userid", how = "left") # Left Join

userid    name        msg
0       1      AJ        hmm
1       1      AJ       acha
2       2   Manoj  theek hai
3       3  Mukesh        NaN

In [125]:
users.merge(msgs, on ="userid", how = "right") # Right Join

userid   name        msg
0       1     AJ        hmm
1       1     AJ       acha
2       2  Manoj  theek hai
3       4    NaN       nice

**What to do when the column names of the 2 DataFrames are not the same ???**    
**How can we merge DataFrames in such a case ???**

By explicitly specifying the columns to merge on using **left_on** and **right_on** parameters, you can handle scenarios where the column names in the DataFrames are not the same but there is a common column or index to merge on.






In [126]:
users.rename(columns = {"userid" : "id"} , inplace = True)

In [127]:
users

id    name
0   1      AJ
1   2   Manoj
2   3  Mukesh

In [128]:
msgs

userid        msg
0       1        hmm
1       1       acha
2       2  theek hai
3       4       nice

In [129]:
users.merge(msgs, left_on = "id", right_on = "userid")

id   name  userid        msg
0   1     AJ       1        hmm
1   1     AJ       1       acha
2   2  Manoj       2  theek hai

          
**pd.join()** (**`NOTE:`** Additional topic external from **Scaler**)

This function is used to join two DataFrames based on their indices. It is similar to the **merge()** operation but uses the index values of DataFrames for the join operation.

In this example, **df1.join(df2)** performs a join operation on the indices of **df1** and **df2**, resulting in a new DataFrame **joined_df**. The columns from both DataFrames are combined based on the matching indices.



In [130]:
# Create two DataFrames with different column names
df1 = pd.DataFrame({'A': [1, 2, 3],
                    'B': [4, 5, 6]},
                   index=['X', 'Y', 'Z'])
df2 = pd.DataFrame({'C': [7, 8, 9],
                    'D': [10, 11, 12]},
                   index=['X', 'Y', 'Z'])

# Join DataFrames based on index
joined_df = df1.join(df2)

# Print the joined DataFrame
print(joined_df)# Join DataFrames based on index

   A  B  C   D
X  1  4  7  10
Y  2  5  8  11
Z  3  6  9  12


# **Importing Libraries and Dataset "IMDB Director Insights"**
***

In [131]:
import numpy as np
import pandas as pd

In [132]:
!gdown 1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd

Downloading...
From: https://drive.google.com/uc?id=1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
To: /content/movies.csv
100% 112k/112k [00:00<00:00, 13.2MB/s]


In [133]:
!gdown 1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm

Downloading...
From: https://drive.google.com/uc?id=1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm
To: /content/directors.csv
100% 65.4k/65.4k [00:00<00:00, 56.1MB/s]


In [134]:
movies = pd.read_csv("movies.csv")
movies.head()

Unnamed: 0     id     budget  popularity     revenue  \
0           0  43597  237000000         150  2787965087   
1           1  43598  300000000         139   961000000   
2           2  43599  245000000         107   880674609   
3           3  43600  250000000         112  1084939099   
4           5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

### **How to remove / drop an unncessary column called "Unnamed: 0" ???**

There are 2 ways in order to remove such unncessary columns as shown below:

> **`Approach (1):` movies = movies.drop(columns = ["Unnamed: 0"], inplace = True)**

> **`Approach (2):` movies = pd.read_csv('movies.csv', index_col = 0)**

Let's work on both these approaches...

In [135]:
#  Approach (1)

movies.drop(columns = ["Unnamed: 0"], inplace = True)

In [136]:
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

In [137]:
# Downloading again just for understanding both the Approaches
!gdown 1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd

movies = pd.read_csv('movies.csv')

Downloading...
From: https://drive.google.com/uc?id=1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
To: /content/movies.csv
100% 112k/112k [00:00<00:00, 18.1MB/s]


In [138]:
movies.head()

Unnamed: 0     id     budget  popularity     revenue  \
0           0  43597  237000000         150  2787965087   
1           1  43598  300000000         139   961000000   
2           2  43599  245000000         107   880674609   
3           3  43600  250000000         112  1084939099   
4           5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

The parameter **index_col = 0** specifies that the first column in the CSV file should be used as the index column of the DataFrame.

In [139]:
# Approach (2)

movies = pd.read_csv('movies.csv', index_col = 0)

In [140]:
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
5                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
5         4767  2007   May   Tuesday

### **How many Movies data features/columns do I have ???**

In [141]:
movies.shape

(1465, 11)

In [142]:
directors = pd.read_csv('directors.csv', index_col=0)

In [143]:
directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

In [144]:
directors.shape

(2349, 3)

**For every director, how many movies they have directed ???**

In [145]:
movies["director_id"].value_counts()

4799    26
4809    19
5087    19
5457    18
4779    16
        ..
5105     5
4849     5
5096     5
5088     5
6204     5
Name: director_id, Length: 199, dtype: int64

**How many different directors have directed a movie ???**

In [146]:
movies["director_id"].nunique()

199

### **Among the directors present in movies df, how many are present in directors df ???**

As you have noticed before, there are **2349 directors** and we also see that only **199 directors** have directed a movie.

Hence, **among the directors present in movies df, how many are present in directors df ???**

In [147]:
movies.shape

(1465, 11)

In [148]:
directors.shape

(2349, 3)

In [149]:
movies["director_id"]

0       4762
1       4763
2       4764
3       4765
5       4767
        ... 
4736    4809
4743    5369
4748    5148
4749    5535
4768    5097
Name: director_id, Length: 1465, dtype: int64

In [150]:
directors["id"]

0       4762
1       4763
2       4764
3       4765
4       4766
        ... 
2344    7106
2345    7107
2346    7108
2347    7109
2348    7110
Name: id, Length: 2349, dtype: int64

In [151]:
#among the directors present in movies df, how many are present in directors df ???

**How to check if the directors in movies dataset are present in the director dataset ???**

The code **movies["director_id"].isin(directors["id"])** checks if the values in the **"director_id"** column of the DataFrame **movies** are present in the **"id"** column of the DataFrame **directors**. It returns a boolean Series with **True** for the rows where the condition is met and **False** otherwise.

In this example, the boolean Series **is_valid_director** is generated by comparing the "director_id" column of **movies** with the "id" column of **directors**. It indicates which rows in **movies** have a valid director ID present in **directors**. The resulting Series contains **True** for the rows where the condition is satisfied, and **False** otherwise.



In [152]:
movies["director_id"].isin(directors["id"]) # Similar to Subquery in SQL

0       True
1       True
2       True
3       True
5       True
        ... 
4736    True
4743    True
4748    True
4749    True
4768    True
Name: director_id, Length: 1465, dtype: bool

**How to validate that every directors in the movies dataset are present in the director dataset ???**

The code **np.all(movies["director_id"].isin(directors["id"]))** checks if all the values in the "director_id" column of the DataFrame **movies** are present in the "id" column of the DataFrame **directors**. It returns **True** if all the values in the "director_id" column are found in the "id" column, and **False** otherwise.

In this example, **np.all(movies["director_id"].isin(directors["id"]))** checks if all the values in the "director_id" column of **movies** are present in the "id" column of **directors**. Since the result is **True**, it means that all the director IDs in **movies** are found in **directors**.



In [153]:
np.all(movies["director_id"].isin(directors["id"]))

True

### **Let's now join these 2 tables**

In [154]:
movies.head(2)

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday

In [155]:
directors.head(2)

director_name    id gender
0   James Cameron  4762   Male
1  Gore Verbinski  4763   Male

**Since the column names of director is different in Movies & Director tables, how to get this fixed ???**

In [156]:
data = movies.merge(directors, left_on = "director_id", right_on = "id")

In [157]:
data.head(2) # Notice id_x, id_y, & director_id. We need to remove the duplicate columns.

id_x     budget  popularity     revenue    title  vote_average  \
0  43597  237000000         150  2787965087   Avatar           7.2   
1  43622  200000000         100  1845034188  Titanic           7.5   

   vote_count  director_id  year month       day  director_name  id_y gender  
0       11800         4762  2009   Dec  Thursday  James Cameron  4762   Male  
1        7562         4762  1997   Nov   Tuesday  James Cameron  4762   Male

In [158]:
data.drop(["director_id", "id_y"], axis = 1, inplace = True)

In [159]:
data.head(2)

id_x     budget  popularity     revenue    title  vote_average  \
0  43597  237000000         150  2787965087   Avatar           7.2   
1  43622  200000000         100  1845034188  Titanic           7.5   

   vote_count  year month       day  director_name gender  
0       11800  2009   Dec  Thursday  James Cameron   Male  
1        7562  1997   Nov   Tuesday  James Cameron   Male

# **Feature Exploration**
***
**Feature exploration** is an essential step in data analysis and machine learning tasks. It involves gaining a deeper understanding of the features or variables in your dataset to extract insights and make informed decisions during the data analysis process. Here is an example of a common technique for feature exploration:

**Descriptive Statistics:**           
Calculate basic **statistical measures** such as mean, median, standard deviation, minimum, maximum, and quartiles to understand the central tendency, spread, and distribution of each feature.


### **How to check if the data has any null values and by how much ???**

In [160]:
data.info() # Gender column has 1341 non-null. Hence, null value = 1465 - 1341 = 124

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1465 entries, 0 to 1464
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id_x           1465 non-null   int64  
 1   budget         1465 non-null   int64  
 2   popularity     1465 non-null   int64  
 3   revenue        1465 non-null   int64  
 4   title          1465 non-null   object 
 5   vote_average   1465 non-null   float64
 6   vote_count     1465 non-null   int64  
 7   year           1465 non-null   int64  
 8   month          1465 non-null   object 
 9   day            1465 non-null   object 
 10  director_name  1465 non-null   object 
 11  gender         1341 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 148.8+ KB


### **How to find statistical information about the dataFrame ???**

In [161]:
data.describe() # Any numerical data will appear using this function

id_x        budget   popularity       revenue  vote_average  \
count   1465.000000  1.465000e+03  1465.000000  1.465000e+03   1465.000000   
mean   45225.191126  4.802295e+07    30.855973  1.432539e+08      6.368191   
std     1189.096396  4.935541e+07    34.845214  2.064918e+08      0.818033   
min    43597.000000  0.000000e+00     0.000000  0.000000e+00      3.000000   
25%    44236.000000  1.400000e+07    11.000000  1.738013e+07      5.900000   
50%    45022.000000  3.300000e+07    23.000000  7.578164e+07      6.400000   
75%    45990.000000  6.600000e+07    41.000000  1.792469e+08      6.900000   
max    48395.000000  3.800000e+08   724.000000  2.787965e+09      8.300000   

         vote_count         year  
count   1465.000000  1465.000000  
mean    1146.396587  2002.615017  
std     1578.077438     8.680141  
min        1.000000  1976.000000  
25%      216.000000  1998.000000  
50%      571.000000  2004.000000  
75%     1387.000000  2009.000000  
max    13752.000000  2016.000000

### **How to find statistical data about object datatype (i.e. String) ???**

In [162]:
data.describe(include = object)

title month     day     director_name gender
count     1465  1465    1465              1465   1341
unique    1465    12       7               199      2
top     Avatar   Dec  Friday  Steven Spielberg   Male
freq         1   193     654                26   1309

# **Getting back to the original data**
***

In [163]:
data.head()

id_x     budget  popularity     revenue                       title  \
0  43597  237000000         150  2787965087                      Avatar   
1  43622  200000000         100  1845034188                     Titanic   
2  43876  100000000         101   520000000  Terminator 2: Judgment Day   
3  43879  115000000          38   378882411                   True Lies   
4  44184   70000000          24    90000098                   The Abyss   

   vote_average  vote_count  year month        day  director_name gender  
0           7.2       11800  2009   Dec   Thursday  James Cameron   Male  
1           7.5        7562  1997   Nov    Tuesday  James Cameron   Male  
2           7.7        4185  1991   Jul     Monday  James Cameron   Male  
3           6.8        1116  1994   Jul   Thursday  James Cameron   Male  
4           7.1         808  1989   Aug  Wednesday  James Cameron   Male

**As we can see, the "revenue" and "budget" columns have such a huge numbers, let's make it as "million" numbers**

In [164]:
# Let's round it to 2 decimal place

data["revenue"] = (data["revenue"]/1000000).round(2)

In [165]:
data.head()

id_x     budget  popularity  revenue                       title  \
0  43597  237000000         150  2787.97                      Avatar   
1  43622  200000000         100  1845.03                     Titanic   
2  43876  100000000         101   520.00  Terminator 2: Judgment Day   
3  43879  115000000          38   378.88                   True Lies   
4  44184   70000000          24    90.00                   The Abyss   

   vote_average  vote_count  year month        day  director_name gender  
0           7.2       11800  2009   Dec   Thursday  James Cameron   Male  
1           7.5        7562  1997   Nov    Tuesday  James Cameron   Male  
2           7.7        4185  1991   Jul     Monday  James Cameron   Male  
3           6.8        1116  1994   Jul   Thursday  James Cameron   Male  
4           7.1         808  1989   Aug  Wednesday  James Cameron   Male

In [166]:
# HW - budget(M)

In [167]:
data["budget"] = (data["budget"]/1000000).round(2)

In [168]:
data.head()

id_x  budget  popularity  revenue                       title  \
0  43597   237.0         150  2787.97                      Avatar   
1  43622   200.0         100  1845.03                     Titanic   
2  43876   100.0         101   520.00  Terminator 2: Judgment Day   
3  43879   115.0          38   378.88                   True Lies   
4  44184    70.0          24    90.00                   The Abyss   

   vote_average  vote_count  year month        day  director_name gender  
0           7.2       11800  2009   Dec   Thursday  James Cameron   Male  
1           7.5        7562  1997   Nov    Tuesday  James Cameron   Male  
2           7.7        4185  1991   Jul     Monday  James Cameron   Male  
3           6.8        1116  1994   Jul   Thursday  James Cameron   Male  
4           7.1         808  1989   Aug  Wednesday  James Cameron   Male

In [169]:
# select * from movies where vote_avg > 7
# How to do such SQL commands in Pandas ???

### **Data Manipulation:** How to Filter rows based on a condition which is similar to what we use to do in SQL ???

Pandas provides a rich set of functions for manipulating and transforming data. Here are some common operations:

> **Filter rows based on a condition**    
**`Approach (1):`**Ex: **`filtered_df = df[df['column_name'] > 5]`**     
**`Approach (2):`** Ex: **`filtered_df = df.loc['column_name'] > 5]`**

> **Sort the DataFrame by a column**       
Ex: **`sorted_df = df.sort_values('column_name')`** (**`NOTE:`** Check under the topics **`"Give me the top 5 popular movies in my dataset."`**)

> **Merge/Join two DataFrames based on a common column**  
Ex: **`merged_df = pd.merge(df1, df2, on = 'column_name')`** (**`NOTE:`** already covered...)

**Filter rows based on a condition**  

**Approach (1):**    
Using normal mask approach with filter condition.

In [170]:
data["vote_average"] > 7

0        True
1        True
2        True
3       False
4        True
        ...  
1460    False
1461    False
1462    False
1463    False
1464    False
Name: vote_average, Length: 1465, dtype: bool

In [171]:
data[ data["vote_average"] > 7 ]

id_x  budget  popularity  revenue                       title  \
0     43597  237.00         150  2787.97                      Avatar   
1     43622  200.00         100  1845.03                     Titanic   
2     43876  100.00         101   520.00  Terminator 2: Judgment Day   
4     44184   70.00          24    90.00                   The Abyss   
5     46000   18.50          67   183.32                      Aliens   
...     ...     ...         ...      ...                         ...   
1424  47488    4.00          11    35.56       Bowling for Columbine   
1426  48310    0.16           3     6.71                  Roger & Me   
1433  46168   11.50          15    23.24      The Remains of the Day   
1438  47597    0.00          11     0.00                     Maurice   
1441  47232    5.00          35     8.20               The Machinist   

      vote_average  vote_count  year month        day  director_name gender  
0              7.2       11800  2009   Dec   Thursday  James Cameron   Male  
1              7.5        7562  1997   Nov    Tuesday  James Cameron   Male  
2              7.7        4185  1991   Jul     Monday  James Cameron   Male  
4              7.1         808  1989   Aug  Wednesday  James Cameron   Male  
5              7.7        3220  1986   Jul     Friday  James Cameron   Male  
...            ...         ...   ...   ...        ...            ...    ...  
1424           7.3         453  2002   Oct  Wednesday  Michael Moore   Male  
1426           7.4          90  1989   Sep     Friday  Michael Moore   Male  
1433           7.5         202  1993   Nov     Friday    James Ivory   Male  
1438           7.1          61  1987   Sep     Friday    James Ivory   Male  
1441           7.3        1247  2004   Feb     Monday  Brad Anderson   Male  

[301 rows x 12 columns]

**Approach (2):**    
**`NOTE:`** The **loc()** also does the same and it's important to start using this as it's as per company standards

In [172]:
data.loc[ data["vote_average"] > 7 ]

id_x  budget  popularity  revenue                       title  \
0     43597  237.00         150  2787.97                      Avatar   
1     43622  200.00         100  1845.03                     Titanic   
2     43876  100.00         101   520.00  Terminator 2: Judgment Day   
4     44184   70.00          24    90.00                   The Abyss   
5     46000   18.50          67   183.32                      Aliens   
...     ...     ...         ...      ...                         ...   
1424  47488    4.00          11    35.56       Bowling for Columbine   
1426  48310    0.16           3     6.71                  Roger & Me   
1433  46168   11.50          15    23.24      The Remains of the Day   
1438  47597    0.00          11     0.00                     Maurice   
1441  47232    5.00          35     8.20               The Machinist   

      vote_average  vote_count  year month        day  director_name gender  
0              7.2       11800  2009   Dec   Thursday  James Cameron   Male  
1              7.5        7562  1997   Nov    Tuesday  James Cameron   Male  
2              7.7        4185  1991   Jul     Monday  James Cameron   Male  
4              7.1         808  1989   Aug  Wednesday  James Cameron   Male  
5              7.7        3220  1986   Jul     Friday  James Cameron   Male  
...            ...         ...   ...   ...        ...            ...    ...  
1424           7.3         453  2002   Oct  Wednesday  Michael Moore   Male  
1426           7.4          90  1989   Sep     Friday  Michael Moore   Male  
1433           7.5         202  1993   Nov     Friday    James Ivory   Male  
1438           7.1          61  1987   Sep     Friday    James Ivory   Male  
1441           7.3        1247  2004   Feb     Monday  Brad Anderson   Male  

[301 rows x 12 columns]

### **How to filter 2 or more rows based on a condition ???**

**Give me only the title and director name for the movies having vote average greater than 7.**

The code **data.loc[data["vote_average"] > 7, ["title", "director_name"]]** selects rows from the DataFrame **data** where the "vote_average" column is greater than 7. It then selects only the "title" and "director_name" columns for those selected rows.

In [173]:
data.loc[ data["vote_average"] > 7 , ["title", "director_name"]]

title  director_name
0                         Avatar  James Cameron
1                        Titanic  James Cameron
2     Terminator 2: Judgment Day  James Cameron
4                      The Abyss  James Cameron
5                         Aliens  James Cameron
...                          ...            ...
1424       Bowling for Columbine  Michael Moore
1426                  Roger & Me  Michael Moore
1433      The Remains of the Day    James Ivory
1438                     Maurice    James Ivory
1441               The Machinist  Brad Anderson

[301 rows x 2 columns]

### **Data Cleaning:** How to Detect and handle 2 conditions ???   

**Data cleaning** is an essential step in the data analysis process. Pandas offers several methods to handle missing values, duplicates, and outliers. Here are a few common operations:

> **Remove duplicate rows:**    
Ex: **`depulicated_df = df.drop_duplicates()`** (**`NOTE:`** already covered...)

> **Detecting & Handling 2 conditions:**    
**`df = df[(df['column_name'] > lower_bound) & (df['column_name'] < upper_bound)]`**



**Detecting & Handling 2 conditions:**     

**Give me only those columns where the vote_average is greater than 7 and the year is on or after 2015**

The code **data.loc[(data["vote_average"] > 7) & (data["year"] >= 2015)]** demonstrates the usage of two operators, namely the comparison operator **>** and the logical operator **&**.

Here are explanations of these operators and additional commonly used operators in pandas:

**(1) Comparison Operators:**

> **>**: It compares if the value on the left is greater than the value on the right.  
> **<**: Less than operator. It compares if the value on the left is less than the value on the right.      
> **>=**: It compares if the value on the left is greater than or equal to the value on the right.           
> **<=**: It compares if the value on the left is less than or equal to the value on the right.            
> **==**: It compares if the value on the left is equal to the value on the right.   
> **!=**: It compares if the value on the left is not equal to the value on the right.

**(2) Logical Operators:**

> **&**: Element-wise logical AND operator. It performs element-wise logical AND between two conditions, resulting in True if both conditions are True, and False otherwise.  
> **|**: Element-wise logical OR operator. It performs element-wise logical OR between two conditions, resulting in True if at least one condition is True, and False otherwise.      
> **~**: Element-wise logical NOT operator. It negates a condition, resulting in True if the condition is False, and False if the condition is True.  

These operators can be used within the indexing or selection operations of pandas, such as **loc[]** or **iloc[]**, to filter or select specific rows based on conditions.


In [174]:
data.loc[ (data["vote_average"] > 7) & (data["year"] >= 2015) ]

id_x  budget  popularity  revenue                    title  \
93   43867   108.0         167   630.16              The Martian   
166  43641   190.0         102  1506.25                Furious 7   
201  44784    40.0          48   165.48          Bridge of Spies   
361  43724   150.0         434   378.86       Mad Max: Fury Road   
465  43773   135.0         100   532.95             The Revenant   
512  44128    75.0          48   108.15  The Man from U.N.C.L.E.   
639  45301    28.0          57   133.35            The Big Short   
649  44281    44.0          68   155.76        The Hateful Eight   
766  44770    35.0          53   194.56               The Intern   
818  47181     5.0          22    24.80                     Race   
853  45293    28.0          61   201.63   Straight Outta Compton   
861  45194    30.0          65    91.71                 Southpaw   

     vote_average  vote_count  year month        day  \
93            7.6        7268  2015   Sep  Wednesday   
166           7.3        4176  2015   Apr  Wednesday   
201           7.2        2583  2015   Oct   Thursday   
361           7.2        9427  2015   May  Wednesday   
465           7.3        6396  2015   Dec     Friday   
512           7.1        2265  2015   Aug   Thursday   
639           7.3        2607  2015   Dec     Friday   
649           7.6        4274  2015   Dec     Friday   
766           7.1        1881  2015   Sep   Thursday   
818           7.1         478  2016   Feb     Friday   
853           7.7        1355  2015   Aug   Thursday   
861           7.3        2067  2015   Jun     Monday   

                   director_name  gender  
93                  Ridley Scott    Male  
166                    James Wan    Male  
201             Steven Spielberg    Male  
361                George Miller    Male  
465  Alejandro González Iñárritu    Male  
512                  Guy Ritchie    Male  
639                   Adam McKay    Male  
649            Quentin Tarantino    Male  
766                 Nancy Meyers  Female  
818              Stephen Hopkins    Male  
853                 F. Gary Gray    Male  
861                Antoine Fuqua    Male

In [175]:
# | --> or
# & --> and

### **Give me only those movies that was released on "Friday" or "Saturday"**

The code **data[(data["day"] == "Friday") | (data["day"] == "Saturday")]** demonstrates the usage of the comparison operator **==** and the logical operator **|**. It selects rows from the DataFrame **data** where the "day" column is equal to either "Friday" or "Saturday".

In [176]:
data[ (data["day"] == "Friday") | (data["day"] == "Saturday")]

id_x  budget  popularity  revenue  \
5     46000   18.50          67   183.32   
6     47036    6.40          74    78.37   
7     43598  300.00         139   961.00   
16    44041   80.00          49   181.00   
17    44154   72.00          32    96.89   
...     ...     ...         ...      ...   
1455  46574   10.00          33   163.88   
1457  46827    0.00           5     0.19   
1458  47152    4.00          42   152.93   
1462  47524    3.00           5     0.00   
1464  48229    0.25           1     4.19   

                                         title  vote_average  vote_count  \
5                                       Aliens           7.7        3220   
6                               The Terminator           7.3        4128   
7     Pirates of the Caribbean: At World's End           6.9        4500   
16                           Road to Perdition           7.3        1077   
17                                     Jarhead           6.6         765   
...                                        ...           ...         ...   
1455                                   Saw III           6.1        1071   
1457                   Repo! The Genetic Opera           6.7         100   
1458                                    Saw II           6.3        1251   
1462                               Please Give           6.0          57   
1464                          Lovely & Amazing           6.3          23   

      year month       day        director_name  gender  
5     1986   Jul    Friday        James Cameron    Male  
6     1984   Oct    Friday        James Cameron    Male  
7     2007   May  Saturday       Gore Verbinski    Male  
16    2002   Jul    Friday           Sam Mendes    Male  
17    2005   Nov    Friday           Sam Mendes    Male  
...    ...   ...       ...                  ...     ...  
1455  2006   Oct    Friday  Darren Lynn Bousman    Male  
1457  2008   Jul    Friday  Darren Lynn Bousman    Male  
1458  2005   Oct    Friday  Darren Lynn Bousman    Male  
1462  2010   Jan    Friday    Nicole Holofcener  Female  
1464  2001   Aug    Friday    Nicole Holofcener  Female  

[701 rows x 12 columns]

### **Give me the top 5 popular movies in my dataset.**

The code **data.sort_values(["popularity"], ascending=False).head(5)** sorts the DataFrame **data** based on the values in the "popularity" column in descending order (highest to lowest) and returns the top 5 rows with the highest popularity values.




In [177]:
data.sort_values(["popularity"], ascending = False).head(5)

id_x  budget  popularity  revenue  \
23   43692   165.0         724   675.12   
361  43724   150.0         434   378.86   
11   43796   140.0         271   655.01   
315  43797   125.0         206   752.10   
22   43662   185.0         187  1004.56   

                                                 title  vote_average  \
23                                        Interstellar           8.1   
361                                 Mad Max: Fury Road           7.2   
11   Pirates of the Caribbean: The Curse of the Bla...           7.5   
315              The Hunger Games: Mockingjay - Part 1           6.6   
22                                     The Dark Knight           8.2   

     vote_count  year month        day      director_name gender  
23        10867  2014   Nov  Wednesday  Christopher Nolan   Male  
361        9427  2015   May  Wednesday      George Miller   Male  
11         6985  2003   Jul  Wednesday     Gore Verbinski   Male  
315        5584  2014   Nov    Tuesday   Francis Lawrence   Male  
22        12002  2008   Jul  Wednesday  Christopher Nolan   Male

### **Give me all the movies directed by "Christopher Nolan".**

The code **data.loc[data["director_name"] == "Christopher Nolan", ["title"]]** selects rows from the DataFrame **data** where the "director_name" column is equal to "Christopher Nolan". It then selects only the "title" column for those selected rows.



In [178]:
data.loc[data["director_name"] == "Christopher Nolan", ["title"]]

title
21  The Dark Knight Rises
22        The Dark Knight
23           Interstellar
24              Inception
25          Batman Begins
26               Insomnia
27           The Prestige
28                Memento

# Similar to "Vectarization" in NumPy, let's learn "apply()" function in Pandas..

**Similar to "Vectarization" that we had learnt previously in "NumPy", we have something called "apply()" in Pandas that we will learn in-deepth in next class.**

In the case of **vectorization** in **NumPy** and **Pandas**, it involves applying operations to entire arrays or series without explicit loops. This approach is significantly faster and more efficient compared to traditional iterative methods. Vectorized operations take advantage of optimized algorithms and parallel processing capabilities, resulting in improved performance and reduced execution time.



### **How to convert Male & Female category into a numberical values as 0 & 1 respectly ???**

Male --> 0          
Otherwise --> 1

In [180]:
data.head()

id_x  budget  popularity  revenue                       title  \
0  43597   237.0         150  2787.97                      Avatar   
1  43622   200.0         100  1845.03                     Titanic   
2  43876   100.0         101   520.00  Terminator 2: Judgment Day   
3  43879   115.0          38   378.88                   True Lies   
4  44184    70.0          24    90.00                   The Abyss   

   vote_average  vote_count  year month        day  director_name gender  
0           7.2       11800  2009   Dec   Thursday  James Cameron   Male  
1           7.5        7562  1997   Nov    Tuesday  James Cameron   Male  
2           7.7        4185  1991   Jul     Monday  James Cameron   Male  
3           6.8        1116  1994   Jul   Thursday  James Cameron   Male  
4           7.1         808  1989   Aug  Wednesday  James Cameron   Male

In [181]:
def encode(data):
  if data == "Male":
    return 0
  else:
    return 1

In [182]:
data["gender"] = data["gender"].apply(encode)

In [183]:
data.head()

id_x  budget  popularity  revenue                       title  \
0  43597   237.0         150  2787.97                      Avatar   
1  43622   200.0         100  1845.03                     Titanic   
2  43876   100.0         101   520.00  Terminator 2: Judgment Day   
3  43879   115.0          38   378.88                   True Lies   
4  44184    70.0          24    90.00                   The Abyss   

   vote_average  vote_count  year month        day  director_name  gender  
0           7.2       11800  2009   Dec   Thursday  James Cameron       0  
1           7.5        7562  1997   Nov    Tuesday  James Cameron       0  
2           7.7        4185  1991   Jul     Monday  James Cameron       0  
3           6.8        1116  1994   Jul   Thursday  James Cameron       0  
4           7.1         808  1989   Aug  Wednesday  James Cameron       0

# **`NOTE:`** we will cover more examples of **apply()** in next lecture...